### DISCUSSION TITLE
How do I make the bot connect through a proxy and not my IP (IN PYTHON, NOT JS)?

### DISCUSSION THREAD

**Discussion by MooshiMochi** [PrismarineJS/mineflayer#2283] [Q&A]
## Versions
 - mineflayer: ^3.11.2
 - server: vanilla 1.8.9
 - node: 16.7.0

## Detailed description of a problem
A clear and concise description of what the problem is, with as much context as possible.

I want to make the bot log in through a proxy in python using the mineflayer API however, when the bot logs in, it uses my IP address rather than the proxy IP. When I tried the js code, it worked like a charm, but it doesn't seem to work in python.

What are you building? What problem are you trying to solve?

I am building a bot that is able to log into my server through multiple proxies (at my choosing)

## What did you try yet?

I have tried to re-write the code exactly as it looks like in javascript in python: 
```py

## THIS IS PYTHON CODE ---------------------- DOESN'T CONNECT THROUGH PROXY, BUT MY IP
bot = mineflayer.createBot({
    'host': 'mooshi_.aternos.me',
    'port': 21615,
    'username': "my_mojang_email",  # email
    "password": "my_mojang_password",  # password
    "auth": 'mojang',  # place a # infront if it is not a migrated account
    "connect: bot =>": socks.createConnection({
            "proxy": {
                "host": "208.121.131.150",
                "port": 9279,
                "type": 5
            },
            "command": "connect",
            "destination": {
                "host": "mooshi_.aternos.me",
                "port": 25565
            }
        }, {"info => ": "bot.setSocket(info.socket) bot.emit('connect')"}
            )
})
```
```js
// THIS IS JAVASCRIPT CODE ------------------- IT WORKS AS I WANT IT TO
const bot = mineflayer.createBot({
  connect: bot => {
    socks.createConnection({
      proxy: {
        host: proxyHost,
        port: parseInt(proxyPort),
        type: 5
      },
      command: 'connect',
      destination: {
        host: server_host,
        port: parseInt(server_port)
      }
    }, (err, info) => {
      if (err) {
        console.log(err)
        return
      }

      bot.setSocket(info.socket)
      bot.emit('connect')
    })
  },
  auth: "mojang",
  username: username,
  password: password,
  host: server_host,
  port: parseInt(server_port)
});
```
Did you try any method from the API?
n/a

Did you try any example? Any error from those?
I tried the javascript example inside a javascript file --> No errors, connects through the proxy as intended.
I tried writing the code in python, --> Bot was able to log in / listen to events, etc however it logged in with my IP instead of the proxy IP
I would have tried an example from the python examples section in the git, but there are no examples that tell you how to connect the bot through a proxy.

## Your current code
```py

## THIS IS PYTHON CODE -----------

import time
from javascript import require, On

mineflayer = require('mineflayer')
socks = require('socks').SocksClient

bot = mineflayer.createBot({
    'host': 'mooshi_.aternos.me',
    'port': 21615,
    'username': "my_mojang_email",  # email
    "password": "my_mojang_passwprd",  # password
    "auth": 'mojang',  # place a # infront if it is not a migrated account
    "connect: bot =>": socks.createConnection({
            "proxy": {
                "host": "208.121.131.150",
                "port": 9279,
                "type": 5
            },
            "command": "connect",
            "destination": {
                "host": "mooshi_.aternos.me",
                "port": 25565
            }
        }, {"function(info) ": "bot.setSocket(info.socket) bot.emit('connect')"}
            )
})

print("Started mineflayer")

@On(bot, 'messagestr')
def handleMsg(_, rawmessage, *args):
    pass

@On(bot, "end")
def handle_end(*args):
    print("Bot ended!", args)


@On(bot, "error")
def handle_error(*args):
    print("Bot error!", args)


@On(bot, "kicked")
def handle_kick(*args):
    print("Bot kick!", args)


first_time_spawning = True

@On(bot, "spawn")
def handlespawn(*args):
    global first_time_spawning
    if first_time_spawning:
        bot.chat("Hello MooshiMochi")
        first_time_spawning = False


@On(bot, "windowOpen")
def window(*args):
    time.sleep(1)
    bot.clickWindow(14, 0, 0)
```

## Expected behavior

When the bot logs in, I expect it to show on the server console that the bot has logged in with the proxy IP

## Additional context

It seems like the script completely ignores the part where i create a socks connection.

**Comment by XfedeX** [comment]
I'd suggest you to just use Javascript.
Mineflayer for Python isn't as supported, it has more dependencies to install, and has no advantages, apart from using a slower language that annoys you for indentation.

**Reply by MooshiMochi** [reply]
Ye, but the problem is that js is a "single thread" programming language, aka I'd doesn't exactly support concurrency apart from asynchronous functions. This is a problem because I have to think of ways to write code without using while loops or any elongated for loops, which is not the case for python at all (you can just run the loop in a different thread) hence y I asked about how to do it in python through a proxy.

**Comment by rom1504** [comment]
https://github.com/PrismarineJS/mineflayer/discussions/2277 seems to be a duplicate
